File: python/calibration.py
#!/usr/bin/python3
import yaml
from mpu9250_driver.mpu9250 import MPU9250
from utils.orientation import quat2eul, eul2quat
import time
import numpy as np
import math

if __name__ == '__main__':
    """
    Do calibration for gyroscope and accelerometer
    """
    nav_frame = "NED" # ENU/NED
    axis = 9
    hz = 100
    interval = 1/hz
    acc_time = 1
    gyro_time = 1
    mag_time = 10
    calibration = True
    imu = MPU9250(nav_frame, axis, hz, calibration)
    imu.initialization()
    gyro_scale, gyro_bias, gyro_misalignment = imu.mpu6500.gyro_calibration(gyro_time)
    accel_scale, accel_bias, accel_misalignment = imu.mpu6500.accel_calibration(acc_time)
    mag_scale, mag_bias, mag_misalignment, mag_strength = imu.ak8963.mag_calibration(mag_time)
    bias = np.vstack((np.vstack((gyro_bias,accel_bias)),mag_bias))
    scale = np.vstack((np.vstack((gyro_scale,accel_scale)),mag_scale))
    misalignment = np.vstack((np.vstack((gyro_misalignment,accel_misalignment)),mag_misalignment))

    # Refresh new config to yaml file
    config = yaml.load(open("../cfg/config.yaml", "r"), Loader=yaml.FullLoader)
    bias_parameter = ["gx_bias","gy_bias","gz_bias","ax_bias","ay_bias","az_bias","mx_bias","my_bias","mz_bias"]
    scale_parameter = ["gx_scale","gy_scale","gz_scale","ax_scale","ay_scale","az_scale","mx_scale","my_scale","mz_scale"]
    
    bias_scale = {}
    for i, element in enumerate(bias_parameter):
        bias_scale[element] = float(bias[i][0])
    for i, element in enumerate(scale_parameter):
        bias_scale[element] = float(scale[i][0])
    config[nav_frame] = bias_scale
    with open("config.yaml", 'w') as file:
        file.write(yaml.dump(config))
    print("accel bias: ")
    print(accel_bias)
    print("gyro bias: ")
    print(gyro_bias)
    print("mag bias: ")
    print(mag_bias)
    print("accel scale: ")
    print(accel_scale)
    print("gyro scale: ")
    print(gyro_scale)
    print("mag scale: ")
    print(mag_scale)

File: cpp/src/calibration.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <eigen3/Eigen/Dense>
#include <yaml-cpp/yaml.h>
#include "../lib/utils.h"
#include "../lib/orientation.h"
#include "../lib/transformation.h"
#include "../lib/mpu9250.h"

using namespace std;
using namespace Eigen;
using namespace YAML;

int main(int argc, char **argv)
{
    string nav_frame = "NED"; // ENU/NED
    int axis = 9;
    float hz = 100;
    float acc_time = 1;
    float gyro_time = 1;
    float mag_time = 10;
    bool calibration = true;
    accel_gyro_calib accel_result, gyro_result;
    mag_calib mag_result;
    float mag_strength;
    Vector3d accel_scale, accel_bias, gyro_scale, gyro_bias, mag_scale, mag_bias;
    VectorXd bias = VectorXd(9);
    VectorXd scale = VectorXd(9);
    VectorXd accel_misalignment = VectorXd(6);
    VectorXd gyro_misalignment = VectorXd(6);
    VectorXd mag_misalignment = VectorXd(6);

    MPU9250 mpu9250(nav_frame, axis, hz, calibration);
    mpu9250.initialization();
    gyro_result = mpu9250.mpu6500.gyro_calibration(gyro_time);
    gyro_scale = gyro_result.scale;
    gyro_bias = gyro_result.bias;
    gyro_misalignment = gyro_result.misalignment;
    cout << "gyro_scale:" << endl;
    cout << gyro_scale << endl;
    cout << "gyro_bias:" << endl;
    cout << gyro_bias << endl;
    cout << "gyro_misalignment:" << endl;
    cout << gyro_misalignment << endl;

    accel_result = mpu9250.mpu6500.accel_calibration(acc_time);
    accel_scale = accel_result.scale;
    accel_bias = accel_result.bias;
    accel_misalignment = accel_result.misalignment;
    cout << "accel_scale:" << endl;
    cout << accel_scale << endl;
    cout << "accel_bias:" << endl;
    cout << accel_bias << endl;
    cout << "accel_misalignment:" << endl;
    cout << accel_misalignment << endl;
    
    mag_result = mpu9250.ak8963.mag_calibration(mag_time);
    mag_scale = mag_result.scale;
    mag_bias = mag_result.bias;
    mag_misalignment = mag_result.misalignment;
    mag_strength = mag_result.strength;
    cout << "mag_scale:" << endl;
    cout << mag_scale << endl;
    cout << "mag_bias:" << endl;
    cout << mag_bias << endl;
    cout << "mag_misalignment:" << endl;
    cout << mag_misalignment << endl;
    cout << "mag_strength:" << endl;
    cout << mag_strength << endl;

    bias << gyro_bias, accel_bias, mag_bias;
    scale << gyro_scale, accel_scale, mag_scale;

    // Refresh new config to yaml file
    Node config = LoadFile("../../cfg/config.yaml");
    vector<string> bias_parameter = {"gx_bias", "gy_bias", "gz_bias", "ax_bias", "ay_bias", "az_bias", "mx_bias", "my_bias", "mz_bias"};
    vector<string> scale_parameter = {"gx_scale", "gy_scale", "gz_scale", "ax_scale", "ay_scale", "az_scale", "mx_scale", "my_scale", "mz_scale"};
    map<string, float> bias_scale;

    for (int i=0; i < bias_parameter.size(); i++) {
        bias_scale[bias_parameter[i]] = bias[i];
    }
    for (int i=0; i < scale_parameter.size(); i++) {
        bias_scale[scale_parameter[i]] = scale[i];
    }

    config[nav_frame] = bias_scale;
    ofstream fout("../../cfg/config.yaml");
    fout << config;

    return 0;
}